Diagnostic test session

Nik-09 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

The PR adds the UI to give the diagnostic test from the diagnostic test player page. And, based on the performance in the test, 0, or 1, or 2 topics were recommended.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

**Updated screenshots**
![Screenshot (11)](https://user-images.githubusercontent.com/57531197/202785398-b03a4d55-e5c0-4f2c-a267-e72166d9cfb3.png)
![Screenshot from 2022-11-22 14-45-35](https://user-images.githubusercontent.com/57531197/203291271-8383c6fc-8de0-4a56-8cb1-8f046d43765b.png)
![Screenshot from 2022-11-22 14-37-31](https://user-images.githubusercontent.com/57531197/203291277-03188e59-bc4e-4ede-bad0-659a38071b00.png)
![Screenshot from 2022-11-22 14-32-34](https://user-images.githubusercontent.com/57531197/203291281-1b7f66b3-27f0-47c1-9dd8-4108213d4186.png)
![Screenshot from 2022-11-22 14-16-24](https://user-images.githubusercontent.com/57531197/203291287-8cfb8f3a-3e6f-4b2b-accc-8dc2442c8501.png)



https://user-images.githubusercontent.com/57531197/201539999-7cde1f1b-dcab-4d34-926f-11742078a98b.mp4

https://user-images.githubusercontent.com/57531197/201539996-05011a97-3a4a-4fbb-93ad-213f00fe6c91.mp4

**Existing exploration editor and player**

https://user-images.githubusercontent.com/57531197/202913675-714f9f35-09d2-46d8-bb33-44984e7561e4.mp4

**Existing practice session**

https://user-images.githubusercontent.com/57531197/202913685-9451a8d7-cb97-41c3-997c-1348df929bfb.mp4


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Nik-09 committed:
Adds diagnostic test player page

Nik-09 committed:
Adds html text

Nik-09 committed:
Merge branch 'diagnostic_player' of https://github.com/Nik-09/oppia into diagnostic_player

Nik-09 committed:
Adds frontend unit tests

Nik-09 committed:
Adds backend tests

Nik-09 committed:
Fixes linter issue

Nik-09 committed:
Fixes mypy issues

Nik-09 committed:
Fixes typescript checks

Nik-09 committed:
Updates copyright text.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_player

Nik-09 committed:
Adds diagnostic test model

Nik-09 committed:
Alignes diagnostic test player with the question player

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_player

Nik-09 committed:
Internationalized diagnostic test texts

Nik-09 committed:
Adds css rules

Nik-09 committed:
Adds Qqq text.

Nik-09 committed:
Adds frontend unit tests

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Adds test file to the backend shard.

Nik-09 committed:
Adds unit test

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Fixes typescript issue.

Nik-09 committed:
Fixes merge conflict with the diagnostic player

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_player

Nik-09 committed:
Sort i18n keys

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_algorithm

Nik-09 committed:
Adds frontend unit test for the diagnostic test model.

Nik-09 committed:
Merge branch 'diagnostic_player' of https://github.com/Nik-09/oppia into diagnostic_test_algorithm

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_algorithm

Nik-09 committed:
Create diagnostic test topic state model

Nik-09 committed:
Adds submit and init metods for engine service

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_player

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Updates file docstring + removes extra new line

Nik-09 committed:
Merge branch 'diagnostic_player' of https://github.com/Nik-09/oppia into diagnostic_test_algorithm

Nik-09 committed:
Adds frontend unit test

Nik-09 committed:
Replaces number with a named constant.

Nik-09 committed:
Updates UI for the diagnostic page

Nik-09 committed:
Removes redundant css class

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_player

Nik-09 committed:
Loads first card

Nik-09 committed:
Merge branch 'diagnostic_player' of https://github.com/Nik-09/oppia into diagnostic_test_algorithm

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Adds code comment

Nik-09 committed:
Adds diagnostic test topic state model

Nik-09 committed:
Adds diagnostic test topic state model

Nik-09 committed:
Fixes lint issues.

Nik-09 committed:
Adds code comments and unit tests

Nik-09 committed:
Renamed variables

Nik-09 committed:
Adds code comment

Nik-09 committed:
Adds diagnostic test questions handler

Nik-09 committed:
Adds diagnostic test questions API method

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_algorithm

Nik-09 committed:
Renames into diagnostic test topic tracker model

Nik-09 committed:
Updates variable name

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Updates diagnostic test current topic status model

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_topic_state_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_algorithm

Nik-09 committed:
Updates code comments

Nik-09 committed:
Removes type safety check for pop method

Nik-09 committed:
Updates API data schema for the diagnostic test skill id to questions

Nik-09 committed:
Fixes mypy checks

Nik-09 committed:
Fixes typescript check

Nik-09 committed:
Fixes lint issue

Nik-09 committed:
Fixes backend test

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_question_handler

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Fixes merge conflict with the diagnostic test topic tracker branch

Nik-09 committed:
Merge branch 'diagnostic_test_topic_state_model' of https://github.com/Nik-09/oppia into diagnostic_test_engine_service

Nik-09 committed:
Removes unwanted files

Nik-09 committed:
Fixes merge conflict with the questions handler branch

Nik-09 committed:
Adds submit method

Nik-09 committed:
Fixes submit button issue

Nik-09 committed:
Adds logic in the conversation skin component

Nik-09 committed:
Adds test status service

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_question_handler

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Fixes lint issue.

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_topic_state_model

Nik-09 committed:
Adds test for the question player status service

Nik-09 committed:
Fixes merge conflict wrt current topic branch

Nik-09 committed:
Merge branch 'diagnostic_test_question_handler' of https://github.com/Nik-09/oppia into diagnostic_test_engine_service

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Fixes mypy issues

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Fixes merge conflict wrt question handler branch

Nik-09 committed:
Adds frontend tests

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_engine_service

Nik-09 committed:
Fixes lint and typescript checks

Nik-09 committed:
Removes print stamt

Nik-09 committed:
Fixes lint and typescript checks

Nik-09 committed:
Removes redundant code

Nik-09 committed:
Adds questions for testing

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_session

Nik-09 committed:
Adds customizations

Nik-09 committed:
Adds guidance tile on the math classroom page.

Nik-09 committed:
Adds diagnostic test result page UI

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_question_handler

Nik-09 committed:
Updates handler logic to avoid duplicate questions

Nik-09 committed:
Merge branch 'diagnostic_test_question_handler' of https://github.com/Nik-09/oppia into diagnostic_test_engine_service

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Adds progress functionality.

Nik-09 committed:
Updates backend logic to avoid duplicate questions

Nik-09 committed:
Adds backend unit test for coverage

Nik-09 committed:
Fixes typo

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Adds progress bar

Nik-09 committed:
Updates min question requirement for the diagnostic test skill

Nik-09 committed:
Fixes frontend tests

Nik-09 committed:
Fixes e2e test

Nik-09 committed:
Updates code comments

Nik-09 committed:
Merge branch 'diagnostic_test_question_handler' of https://github.com/Nik-09/oppia into diagnostic_test_engine_service

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_engine_service

Nik-09 committed:
Merge branch 'diagnostic_test_engine_service' of https://github.com/Nik-09/oppia into diagnostic_test_session

Nik-09 committed:
Fixes UI

Nik-09 committed:
Adds skip functionality

Nik-09 committed:
Fixes dummy data creation

Nik-09 committed:
Fixes first topic button

Nik-09 committed:
Fixes result page buttons

Nik-09 committed:
Adds back button

Nik-09 committed:
Adds page unload events

Nik-09 committed:
Fixes console error

Nik-09 committed:
Fixes merge conflict

Nik-09 commented on core/controllers/admin.py:
```diff
@@ -831,18 +831,112 @@ def _generate_dummy_classroom(self) -> None:
             skill_id_5 = skill_services.get_new_skill_id()
             skill_id_6 = skill_services.get_new_skill_id()
 
+            question_id_1 = question_services.get_new_question_id()
```
The changes in this file are used to generate the dummy classrooms with data.

seanlip replied:
Ack. Given these changes, though, shouldn't the admin page text be updated with an explanation of what the generated classroom will contain?

Nik-09 replied:
Updated the text in the admin config tab.
Text: 
`Generate a dummy math classroom with 5 published topics containing 1 skill per topic and 3 questions linked per skill.`

Nik-09 commented on core/controllers/admin.py:
```diff
@@ -857,35 +951,79 @@ def _generate_dummy_classroom(self) -> None:
             skill_services.save_new_skill(self.user_id, skill_5)
             skill_services.save_new_skill(self.user_id, skill_6)
 
-            topic_1.add_uncategorized_skill_id(skill_id_1)
```
Removing uncategorized skills as the skills are now linked with the subtopics.

Nik-09 commented on core/controllers/classroom.py:
```diff
@@ -244,10 +242,10 @@ class ClassroomHandler(
     }
 
     @acl_decorators.can_access_admin_page
-    def get(self, classroom_id: str) -> None:
```
Changing the fetch type from classroom ID to classroom URL fragment.

Nik-09 committed:
Removes unwanted changes

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Fixes mypy checks

Nik-09 committed:
Fixes typescript checks

Nik-09 commented on core/templates/pages/classroom-page/classroom-page.component.html:
```diff
@@ -22,6 +22,31 @@ <h2>{{ 'I18N_CLASSROOM_PAGE_COURSE_DETAILS' | translate }}</h2>
+      </a>
+    </div>
+    <div class="classroom-page-guidance-tile">
+      <p><strong>{{ 'I18N_CLASSROOM_PAGE_ALREADY_KNOW_SOME_MATH_HEADING' | translate}}</strong></p>
```
Make on different line

Nik-09 commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-current-topic-status.model.ts:
```diff
@@ -108,4 +111,8 @@ export class DiagnosticTestCurrentTopicStatusModel {
     return this._pendingSkillIdsToTest.length === 0;
   }
+
+  isLifelineConsumeed(): boolean {
```
This should be updated after merging the base PR

Nik-09 commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player-page.module.ts:
```diff
@@ -38,6 +38,9 @@ import { MyHammerConfig, toastrConfig } from 'pages/oppia-root/app.module';
 import { AppErrorHandlerProvider } from 'pages/oppia-root/app-error-handler';
 import { DiagnosticTestPlayerComponent } from './diagnostic-test-player.component';
+import { InteractiveTextInputComponent } from '../../../../extensions/interactions/TextInput/directives/oppia-interactive-text-input.component';
+import { ResponseTextInputComponent } from '../../../../extensions/interactions/TextInput/directives/oppia-response-text-input.component';
```
Try to relative location path

Nik-09 commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.html:
```diff
@@ -14,21 +26,203 @@
 </div>
 
 <style>
+  .diagnostic-test-page-header {
```
CSS class should prefixed with 'oppia-'

Nik-09 commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.ts:
```diff
@@ -29,18 +33,78 @@ import { PreventPageUnloadEventService } from 'services/prevent-page-unload-even
+  recommendedTopicSummaries: CreatorTopicSummary[] = [];
+  recommendedTopicIds: string[] = [];
+  progressPercentage: number = 0;
+
```
remove new line

Nik-09 commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.ts:
```diff
@@ -29,18 +33,78 @@ import { PreventPageUnloadEventService } from 'services/prevent-page-unload-even
   ngOnInit(): void {
-    this.preventPageUnloadEventService.addListener();
+    this.preventPageUnloadEventService.addListener(() => {
+      return !(!this.diagnosticTestIsStarted || this.diagnosticTestIsFinished);
```
Make this boolean logic simple

Nik-09 commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.ts:
```diff
@@ -29,18 +33,78 @@ import { PreventPageUnloadEventService } from 'services/prevent-page-unload-even
+    return 'Start ' + topicName;
+  }
+
+  getTopicUrlFromUrlFragment(urlFragment: string): string {
```
Add test for this

Nik-09 commented on core/templates/pages/exploration-player-page/layout-directives/progress-nav.component.ts:
```diff
@@ -110,6 +116,12 @@ export class ProgressNavComponent {
   ngOnInit(): void {
     this.isIframed = this.urlService.isIframed();
 
+    if (!this.enableNavigationThroughCardHistory) {
```
Add a comment to explain

Nik-09 commented on core/templates/pages/exploration-player-page/layout-directives/progress-nav.component.ts:
```diff
@@ -133,6 +145,10 @@ export class ProgressNavComponent {
     );
   }
 
+  skipQuestionButton(): void {
```
Add test

Nik-09 commented on core/templates/pages/exploration-player-page/layout-directives/progress-nav.component.ts:
```diff
@@ -145,9 +161,12 @@ export class ProgressNavComponent {
       this.displayedCard.getStateName() === null &&
-      !this.explorationPlayerStateService.isInQuestionMode());
+      !this.explorationPlayerStateService.isInQuestionMode() &&
+      !this.explorationPlayerStateService.isInDiagnosticTestPlayerMode());
```
Add a comment

Nik-09 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -412,6 +424,18 @@ export class ConversationSkinComponent {
         this.visitedStateNames.push(firstStateName);
       }
+
+      if (this.diagnosticTestTopicTrackerModel) {
```
Explain this with a comment

Nik-09 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -989,8 +1013,10 @@ export class ConversationSkinComponent {
     this.nextCard = initialCard;
-    this.explorationPlayerStateService.onPlayerStateChange.emit(
-      this.nextCard.getStateName());
+    if (!this.explorationPlayerStateService.isInDiagnosticTestPlayerMode()) {
```
Add comment

Nik-09 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1066,9 +1106,12 @@ export class ConversationSkinComponent {
-    if (!this.explorationPlayerStateService.isInQuestionMode() &&
-      !this.isInPreviewMode &&
-      AppConstants.ENABLE_SOLICIT_ANSWER_DETAILS_FEATURE) {
+    if (
```
Add comment

Nik-09 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1147,6 +1197,10 @@ export class ConversationSkinComponent {
             new Date().getTime() - timeAtServerCall),
           1.0));
 
+        if (this.explorationPlayerStateService.isInDiagnosticTestPlayerMode()) {
```
Refactor this

Nik-09 commented on core/templates/pages/exploration-player-page/learner-experience/tutor-card.component.ts:
```diff
@@ -197,6 +201,12 @@ export class TutorCardComponent {
         }
       )
     );
+
```
Add comment

Nik-09 commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.spec.ts:
```diff
@@ -0,0 +1,902 @@
+    }));
+
+  it(
+    'should be able to finish test if all the eligible topics are tested',
```
Change the word to -- `pending`

Nik-09 commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts:
```diff
@@ -0,0 +1,457 @@
+    );
+  }
+
+  getNextCardAfterSkippingCurrentQuestion(
```
Make the name to `skipCurrentQuestion()`

Nik-09 commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts:
```diff
@@ -0,0 +1,457 @@
+    let currentTopicIsCompletelyTested = (
+      this.diagnosticTestCurrentTopicStatusModel.isTopicCompletelyTested());
+
+    if (currentTopicIsCompletelyTested) {
```
Use existing method

Nik-09 committed:
Fixes mcq questions

Nik-09 committed:
Adds different module components

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Removes unwanted lines

Nik-09 committed:
Adds handler to get classroom ID

Nik-09 reviewed (commented):
Hi @seanlip, the PR is ready for review.
I am currently adding/fixing some frontend tests, but can you please provide feedback on other parts of the PR?
Thank you

seanlip commented:
Thanks, this looks pretty good from a UI perspective! Some general notes:

- "Take a 10-15 question quiz to ..." (question, not questions)
- Could you please get design review for your videos? Check with @dchen97 and @findimue (or email design@oppia.org).
- In the second video, something seems off with the text of the questions, I suspect text from the previous question is not getting deleted properly. Could you PTAL at that?

seanlip commented on assets/i18n/en.json:
```diff
@@ -163,11 +163,17 @@
     "I18N_CLASSROOM_CALLOUT_HEADING_2": "Introducing: The Oppia Classroom",
     "I18N_CLASSROOM_CALLOUT_PARAGRAPH_1": "Check out the first comprehensive course in the all-new Oppia Classroom! Curated lessons - reviewed by educators - so that you can master basic math skills in topics ranging from Place Values to Multiplication and Division.",
     "I18N_CLASSROOM_MATH_TITLE": "Math",
+	"I18N_CLASSROOM_PAGE_ALREADY_KNOW_SOME_MATH_HEADING": "Already know some math?",
```
Something seems off about the indentation here -- please check the "Files Changed" tab of your PR. Ditto for all the other newly-added lines.

Nik-09 replied:
![image](https://user-images.githubusercontent.com/57531197/201913314-2970a8da-e767-4206-b1c2-95e45e0558b2.png)

Yes, this seems a bit off. But in the local changes, the indentation is correct.

seanlip replied:
You will still need to fix it on GitHub, because that is what counts, and we can't merge this until that is fixed.

Perhaps check if you are using tabs vs spaces.

Nik-09 replied:
Updated

seanlip commented on assets/i18n/en.json:
```diff
@@ -163,11 +163,17 @@
     "I18N_CLASSROOM_CALLOUT_PARAGRAPH_1": "Check out the first comprehensive course in the all-new Oppia Classroom! Curated lessons - reviewed by educators - so that you can master basic math skills in topics ranging from Place Values to Multiplication and Division.",
     "I18N_CLASSROOM_MATH_TITLE": "Math",
+	"I18N_CLASSROOM_PAGE_ALREADY_KNOW_SOME_MATH_HEADING": "Already know some math?",
+	"118N_CLASSROOM_PAGE_ALREADY_KNOW_SOME_MATH_TEXT": "Take a 10-15 questions quiz to find out where to start.",
```
questions --> question

Nik-09 replied:
Done

seanlip commented on assets/i18n/en.json:
```diff
@@ -163,11 +163,17 @@
     "I18N_CLASSROOM_PAGE_COURSE_DETAILS": "Course Details",
     "I18N_CLASSROOM_PAGE_HEADING": "The Oppia Classroom",
+    "I18N_CLASSROOM_PAGE_NEW_TO_MATH_HEADING": "New to math?",
+	"I18N_CLASSROOM_PAGE_NEW_TO_MATH_TEXT": "Start from the basic with out first topic, <[firstTopic]>.",
```
Start from the basics with our ...

Nik-09 replied:
Updated

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -163,11 +163,17 @@
 	"I18N_CLASSROOM_CALLOUT_HEADING_2": "The subheading for the callout to the classroom page which is present in the library page.",
 	"I18N_CLASSROOM_CALLOUT_PARAGRAPH_1": "Text for description of the callout to the classroom page which is present in the library page.",
 	"I18N_CLASSROOM_MATH_TITLE": "The classroom name that is displayed on the classroom page.",
+	"I18N_CLASSROOM_PAGE_ALREADY_KNOW_SOME_MATH_HEADING": "The heading text is on the right tile of the math classroom page. It asks the learner, whether they already know some math.",
```
Drop the comma after "learner".

Nik-09 replied:
Updated

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -163,11 +163,17 @@
 	"I18N_CLASSROOM_CALLOUT_PARAGRAPH_1": "Text for description of the callout to the classroom page which is present in the library page.",
 	"I18N_CLASSROOM_MATH_TITLE": "The classroom name that is displayed on the classroom page.",
+	"I18N_CLASSROOM_PAGE_ALREADY_KNOW_SOME_MATH_HEADING": "The heading text is on the right tile of the math classroom page. It asks the learner, whether they already know some math.",
+	"118N_CLASSROOM_PAGE_ALREADY_KNOW_SOME_MATH_TEXT": "The text guides the learner to take a quiz that contains 10-15 questions. The result of the quiz will help to find out the topic from where the learner can start learning.",
```
Explain where this is displayed. Ditto for all the other descriptions.

Nik-09 replied:
Updated

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -163,11 +163,17 @@
 	"I18N_CLASSROOM_PAGE_COURSE_DETAILS": "The header displayed above the course details section in the classroom page.",
 	"I18N_CLASSROOM_PAGE_HEADING": "The header displayed above the classroom name in the classroom page.",
+	"I18N_CLASSROOM_PAGE_NEW_TO_MATH_HEADING": "The heading text is on the left tile of the math classroom page. It asks the learner, whether they are new to math.",
+	"I18N_CLASSROOM_PAGE_NEW_TO_MATH_TEXT": "The text guides the learner to start with the basic math with our first topic.",
```
Explain where this is displayed.

Nik-09 replied:
Updated

seanlip commented on core/controllers/classroom.py:
```diff
@@ -304,3 +304,27 @@ def get(self, classroom_url_fragment: str) -> None:
+            classroom_url_fragment)
+        if classroom is None:
+            raise self.PageNotFoundException(
+                'The classroom with the given id or url doesn\'t exist.')
```
Why do we say "given ID" when there's no ID given?

Nik-09 replied:
Updated

seanlip commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-current-topic-status.model.ts:
```diff
@@ -87,7 +88,7 @@ export class DiagnosticTestCurrentTopicStatusModel {
 
   getNextSkill(): string {
-    return this._pendingSkillIdsToTest.shift() || '';
+    return this._pendingSkillIdsToTest[0];
```
Should you be error-handling the case where the list is empty?

Nik-09 replied:
Updated

Nik-09 replied:
This method is now removed

seanlip commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.ts:
```diff
@@ -29,19 +33,82 @@ import { PreventPageUnloadEventService } from 'services/prevent-page-unload-even
+  }
+
+  getTopicButtonText(topicName: string): string {
+    return 'Start ' + topicName;
```
This should be translated/internationalized.

Nik-09 replied:
Done

seanlip commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.ts:
```diff
@@ -29,19 +33,82 @@ import { PreventPageUnloadEventService } from 'services/prevent-page-unload-even
+  }
+
+  getTopicUrlFromUrlFragment(urlFragment: string): string {
+    return '/learn/math/' + urlFragment;
```
Use UrlInterpolationService.

Nik-09 replied:
Done

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1130,10 +1175,15 @@ export class ConversationSkinComponent {
             nextCard.getStateName());
-        } else {
+        } else if (
+          !this.explorationPlayerStateService.isInDiagnosticTestPlayerMode()
```
I think these if conditions are getting too confusing.

Can we do something like this -- at init(), determine what mode this is in and store it in a variable that never gets edited again. Then all future decisions on what to do reference that variable.

Nik-09 replied:
Yes, Sean, I understand these “If” conditions may cause confusion.
Currently, I am a little worried to refactor this in the current PR, because of the following reasons:
- The conversation skin component currently handles 3 different engines i.e., the exploration player engine, the question player engine, and the diagnostic test player engine.
- The refactor may cause unforeseen errors which can lead to delaying the current PR and project.
- I can create a new PR but for that, I have to take a look into the exploration-player engine service and how it is connected with the conversation skin component.

In my understanding adding explanatory comments above the “If” conditions may help.

Looking forward to your suggestion.

Thank you

seanlip replied:
I think my concern is that I'm not convinced your code is correct, and it sounds like you're not convinced either since you haven't really looked into the exploration-player engine service. It also makes this whole thing into a "big ball of mud" since folks will be afraid to touch it.

Given this, I would still suggest doing the refactor. My suggestion would be to first do the refactor without the diagnostic test stuff -- just preserve the existing behaviour. Then add the diagnostic test stuff to it -- it will be simpler to do that after the refactor.

Nik-09 replied:
Updated the logic by putting the modes that are expected in a particular block of code. Please take a look.

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1147,6 +1197,10 @@ export class ConversationSkinComponent {
             new Date().getTime() - timeAtServerCall),
           1.0));
 
+        if (this.explorationPlayerStateService.isInDiagnosticTestPlayerMode()) {
```
This should be an if/else with the above since they're mutually exclusive.

Why are we special-casing this, btw?

Nik-09 replied:
Refactored to `if - else if - else` block.

seanlip reviewed (changes requested):
@Nik-09 took a full code pass, too!

Nik-09 committed:
Adds frontend tests

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Updates mili sec to wait condition

Nik-09 committed:
Fixes lint issues

Nik-09 reviewed (commented):
Addressed the comments, currently fixing the unit tests.

Nik-09 reviewed (commented):
Hi, @seanlip I need a suggestion from your side, please take a look.
https://github.com/oppia/oppia/pull/16489#discussion_r1022716552

Thank you

seanlip commented on assets/i18n/en.json:
```diff
@@ -164,13 +164,13 @@
     "I18N_CLASSROOM_PAGE_HEADING": "The Oppia Classroom",
     "I18N_CLASSROOM_PAGE_NEW_TO_MATH_HEADING": "New to math?",
-	"I18N_CLASSROOM_PAGE_NEW_TO_MATH_TEXT": "Start from the basic with out first topic, <[firstTopic]>.",
+	"I18N_CLASSROOM_PAGE_NEW_TO_MATH_TEXT": "Start from the basics with out first topic, <[firstTopic]>.",
```
"out" should be "our". I previously commented on this.

Nik-09 replied:
Updated

seanlip commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-current-topic-status.model.ts:
```diff
@@ -88,7 +88,11 @@ export class DiagnosticTestCurrentTopicStatusModel {
 
   getNextSkill(): string {
-    return this._pendingSkillIdsToTest[0];
+    try {
```
Don't do this. Try/catch is too general and if there are other unexpected errors then you'll (erroneously) catch them when they should be raised.

Handle the empty-list case explicitly. Also, I don't understand why we return '', that is not a valid skill ID. In what circumstances does the empty-list case occur?

Also, the method name should have "skill ID" in it rather than "skill".

Nik-09 replied:
Removed

seanlip commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts:
```diff
@@ -178,6 +178,7 @@ export class DiagnosticTestPlayerEngineService {
         wasOldStateInitial: boolean,
         isFirstHit: boolean,
         isFinalQuestion: boolean,
+        nextCardIfReallyStuck: StateCard,
```
Why is this relevant for the diagnostic test?

Also, for the diagnostic test, if the learner gets it wrong, they just get it wrong, right? We don't show feedback or hints or anything, we just advance. Is that correct?

Nik-09 replied:
Currently, the submit method in the conversation skin handles 3 different engines. Each of the engines has its own submit method but since they are bonded to a single submits method in the conversation skin, some non-relevant information also needs to be passed. 

I will create an issue later today as per our discussion in the meeting.

Nik-09 replied:
Created https://github.com/oppia/oppia/issues/16572

seanlip commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.spec.ts:
```diff
@@ -411,11 +411,13 @@ describe('Diagnostic test engine service', () => {
           .createNew('default', '', '', []), 1, 0, 'default_outcome'
       );
-      answerClassificationResult.outcome.labelledAsCorrect = false;
+      answerClassificationResult.outcome.labelledAsCorrect = true;
```
Why did this change?

Nik-09 replied:
Failing on a topic does not progress to the next skill as we have to present the backup question from the same skill. Thus I have changed it into the correct answer so this will mark the skill as passed and move to the next pending skill ID. And as per the spy in line 422, the next fetching of skill ID will trigger a rejection handler.

This test is changed because of a little modification in the method.

seanlip replied:
OK, this sounds like something where you might want to add explanatory comments to the test to explain what your intentions are (similar to the ones that already exist).

Nik-09 replied:
Added

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/tutor-card.component.spec.ts:
```diff
@@ -224,6 +224,49 @@ describe('Tutor card component', () => {
   }));
 
+  it(
+    'should be able to set celebration component flag to based on diagnostic' +
```
I don't understand this test name. Could you explain it more?

Nik-09 replied:
Removed

Nik-09 committed:
Adds backend and frontend test coverage

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Fixes e2e test yaml file

Nik-09 commented on core/controllers/admin_test.py:
```diff
@@ -611,7 +611,7 @@ def test_generate_dummy_classroom_data(self) -> None:
                 'action': 'generate_dummy_classroom'
             }, csrf_token=csrf_token)
         classrooms = classroom_config_services.get_all_classrooms()
-        self.assertEqual(len(classrooms), 2)
```
Now the dummy data generates only one topic.

Nik-09 commented on core/controllers/classroom_test.py:
```diff
@@ -298,7 +298,7 @@ def test_not_able_to_get_classroom_data_when_user_is_not_admin(
 
     def test_get_new_classroom_id(self) -> None:
         self.login(self.CURRICULUM_ADMIN_EMAIL, is_super_admin=True)
-        json_response = self.get_json(feconf.CLASSROOM_ID_HANDLER_URL)
```
Updating the name as per the functionality.

Nik-09 commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-current-topic-status.model.spec.ts:
```diff
@@ -75,20 +75,21 @@ describe('Diagnostic test current topic status model', () => {
     expect(diagnosticTestCurrentTopicStatusModel.getPendingSkillIds())
       .toEqual(['skillID1', 'skillID2', 'skillID3']);
 
-    let currentSkillId = diagnosticTestCurrentTopicStatusModel.getNextSkill();
```
getNextSkill method is removed, hence these lines are not required anymore.

Nik-09 reviewed (commented):
Some notes

Nik-09 committed:
Fixes indentation

Nik-09 reviewed (commented):
Hi, @seanlip I have addressed the comments please take a look once you are free.
Also added @kevintab95 for providing the feedback.

Thank you.

Nik-09 committed:
Updates progress component

ghost commented:
![image](https://user-images.githubusercontent.com/34294292/202505044-9bd9f1d4-6e86-49e7-9e6f-f6c0a00c08f6.png)
In the "New to math?" box, the word "out" should be "our." 
In the "Already know some math?" box, the word "questions" should be "question."

ghost commented:
![image](https://user-images.githubusercontent.com/34294292/202506058-83828e19-79ce-4500-a858-6c4018ccc588.png)
The user has no instructions to tell them to either select an answer or use the CTA to skip the question. It would be a better experience to have a button to Submit the answer and a secondary-styled button to Skip the question. If that's not possible, then clear instructions should be provided.

ghost commented:
![image](https://user-images.githubusercontent.com/34294292/202506832-778ae898-3cba-4082-867a-f4e2a79d917a.png)
The two CTAs, SKIP and SUBMIT, should be different styles, so that the user is encouraged to Submit. Skip should be a secondary style, i.e., same color scheme but with an outline of the color, a white background, text the same blue color.

ghost commented:
![image](https://user-images.githubusercontent.com/34294292/202507671-29157f78-6da7-4e99-babc-ed39e7609af7.png)
It would be helpful to provide a label for the icon on the right side to indicate what its function is. Is it to Edit? To delete? Without a label, it's not clear what clicking that icon does.

ghost commented:
![image](https://user-images.githubusercontent.com/34294292/202508109-4709b396-d6fa-485a-ae13-cf2ea879e1cf.png)
"Based on your answers in the test, we recommend starting with this topic." should be revised to "Based on your answers, we recommend starting with this topic."

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1239,22 +1287,36 @@ export class ConversationSkinComponent {
           }
         }
-        if (!this.explorationPlayerStateService.isInQuestionMode()) {
+        if (this.explorationPlayerStateService.isInExplorationMode() ||
```
I think you should have a new method that's something like isPresentingIsolatedQuestions() that returns true when the player is presenting questions, and false when it's in story. Document this method carefully to explain what the difference between true/false are, and use that method here.

In that method, do something like "if in these mode, return true, else if in these modes, return false, else throw exception". That way, if someone adds a new mode and forgets to update that, it gets caught.

That helps the reader understand the core concept of what we're basing this condition on, rather than "it's mode A or B or C".

Similarly above -- I think something like this will help clear some of the confusion and make the if conditions similar to understand. Also what is "other mode"? That just seems weird and probably needs to be communicated better -- the aim of the constructs I proposed is to avoid having a "miscellaneous category" because no one will have any idea what goes in it.

Nik-09 replied:
Created the method.

For the ”other“ category, I have not seen any practical use case for it. 
Currently, it is only active while initializing the exploration player state service until any mode is set to the service.
I have removed the logic from the submit method to use “other mode”.

seanlip commented:
@Nik-09 Please also file the issue you mentioned above and link it here -- this is something that should actually be done before passing the PR back for review, since the reviewer needs to check the issue to ensure it makes sense. Thanks!

Nik-09 committed:
Adds e2e test for the  diagnostic test

Nik-09 committed:
Fixes frontend test

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Fixes frontend test

kevintab95 reviewed (approved):
Same comments as Sean in the codeowner file. Approving to unblock. Thanks @Nik-09!

Nik-09 commented:
Some E2E tests are failing, I will assign @U8NWXD after fixing them.

Thank you

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Adds translation text

Nik-09 committed:
Fixes typo

Nik-09 committed:
Adds code comment

Nik-09 committed:
Updates progress nav component

Nik-09 commented:
> ![image](https://user-images.githubusercontent.com/34294292/202505044-9bd9f1d4-6e86-49e7-9e6f-f6c0a00c08f6.png) In the "New to math?" box, the word "out" should be "our." In the "Already know some math?" box, the word "questions" should be "question."

Updated

Nik-09 commented:
> ![image](https://user-images.githubusercontent.com/34294292/202506058-83828e19-79ce-4500-a858-6c4018ccc588.png) The user has no instructions to tell them to either select an answer or use the CTA to skip the question. It would be a better experience to have a button to Submit the answer and a secondary-styled button to Skip the question. If that's not possible, then clear instructions should be provided.

https://github.com/oppia/oppia/blob/dcdf7f3cc131f8243240dbae3c6db6717e387f23/extensions/interactions/interaction_specs.json#L629

For the submit button, we have set the specification for not showing the submit button in the interaction for multiple-choice input.

Regarding providing instruction, can you please provide your suggestion, so that I can think along those lines?
I am thinking to provide a text like “The answer will auto-submit on the first choice.”

Nik-09 commented:
> ![image](https://user-images.githubusercontent.com/34294292/202508109-4709b396-d6fa-485a-ae13-cf2ea879e1cf.png) "Based on your answers in the test, we recommend starting with this topic." should be revised to "Based on your answers, we recommend starting with this topic."

Updated

Nik-09 commented:
> ![image](https://user-images.githubusercontent.com/34294292/202506832-778ae898-3cba-4082-867a-f4e2a79d917a.png) The two CTAs, SKIP and SUBMIT, should be different styles, so that the user is encouraged to Submit. Skip should be a secondary style, i.e., same color scheme but with an outline of the color, a white background, text the same blue color.

Updated, I have added a screenshot in the description section of the PR, PTAL.

Nik-09 commented:
> ![image](https://user-images.githubusercontent.com/34294292/202507671-29157f78-6da7-4e99-babc-ed39e7609af7.png) It would be helpful to provide a label for the icon on the right side to indicate what its function is. Is it to Edit? To delete? Without a label, it's not clear what clicking that icon does.

Created an issue, since this can be handled independently. https://github.com/oppia/oppia/issues/16575

Nik-09 reviewed (commented):
Hi @seanlip I have addressed the review comments, please take a look once you are free.
Thank you

Nik-09 committed:
Fixes lint issues

ghost commented:
> > ![image](https://user-images.githubusercontent.com/34294292/202506058-83828e19-79ce-4500-a858-6c4018ccc588.png) The user has no instructions to tell them to either select an answer or use the CTA to skip the question. It would be a better experience to have a button to Submit the answer and a secondary-styled button to Skip the question. If that's not possible, then clear instructions should be provided.
> 
> https://github.com/oppia/oppia/blob/dcdf7f3cc131f8243240dbae3c6db6717e387f23/extensions/interactions/interaction_specs.json#L629
> 
> For the submit button, we have set the specification for not showing the submit button in the interaction for multiple-choice input.
> 
> Regarding providing instruction, can you please provide your suggestion, so that I can think along those lines? I am thinking to provide a text like “The answer will auto-submit on the first choice.”

Maybe UXW could help with this? I'm not sure how to tag in Dana (user ID @danaPJ)
"When you select an answer, the quiz will automatically move to the next question."

seanlip commented:
Hi @findimue, one question about the UI. What are your thoughts on the "Go Back" button in the top left of the `three.mp3` video? It feels a bit out of place to me and I would have expected it to be in the card that's centered in the page, or something like "Back to Classroom" -- we don't use such a pattern elsewhere.

Also @Nik-09 In the `two.mp3` video, the question text still doesn't seem to clear properly when a new question is shown. Have you taken a look into that?

seanlip commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.html:
```diff
@@ -1,54 +1,242 @@
+      </a>
+    </button>
+    <div class="oppia-diagnostic-test-session">
+      Learner Diagnostic Test
```
Shouldn't this be internationalized?

Nik-09 replied:
Done

seanlip commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.html:
```diff
@@ -1,54 +1,242 @@
+        <div class="oppia-current-progress" [ngStyle]="{'width': progressPercentage + '%'}"></div>
+      </div>
+      <div class="oppia-progress-percentage-container">
+        <span><i>Current progress: {{progressPercentage}}%</i></span>
```
Shouldn't this be internationalized?

Nik-09 replied:
Done

seanlip commented on core/templates/pages/diagnostic-test-player-page/diagnostic-test-player.component.html:
```diff
@@ -1,54 +1,242 @@
+        <div class="oppia-classroom-button-container">
+          <button class="btn oppia-goto-classroom-button">
+            <a class="oppia-anchor-tag-under-button" href="/learn/math">
+              Go to classroom
```
Shouldn't this be internationalized?

Nik-09 replied:
Done

seanlip commented on core/templates/pages/exploration-player-page/layout-directives/progress-nav.component.html:
```diff
@@ -3,7 +3,7 @@
   <div class="row-item row-item-previous-card">
     <!-- ng-if on this second wrapping div because ng-if lags on buttons. -->
-    <div *ngIf="hasPrevious && !helpCardHasContinueButton">
+    <div *ngIf="hasPrevious && !helpCardHasContinueButton && enableNavigationThroughCardHistory">
```
Don't start booleans with verbs; perhaps navigationThroughCardHistoryIsEnabled

Nik-09 replied:
Updated

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -168,6 +169,14 @@ export class ConversationSkinComponent {
 
+  // The fields are used to customize the component for the diagnostic player,
+  // question player, and exploration player page.
+  enableFeedback!: boolean;
```
Don't start variable names with verbs. Consider: feedbackIsEnabled, learnerCanOnlyAttemptQuestionOnce, onlyLastInputResponsePairIsShown, navigationThroughCardHistoryIsEnabled, checkpointCelebrationModalIsEnabled

Nik-09 replied:
Updated

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1220,7 +1260,8 @@ export class ConversationSkinComponent {
         this.nextCardIfStuck = nextCardIfReallyStuck;
         if (!this._editorPreviewMode &&
-            !this.explorationPlayerStateService.isInQuestionMode()) {
+            this.explorationPlayerStateService.isPresentingIsolatedQuestions()
```
If it's presenting isolated questions, then how can it be in preview mode?

Nik-09 replied:
Updated

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1241,22 +1282,36 @@ export class ConversationSkinComponent {
             nextCard.getStateName());
-        } else {
+        } else if (
+          this.explorationPlayerStateService.isInQuestionPlayerMode()
```
If question player mode implies that it's presenting isolated questions, then how can this branch ever execute (given the previous if clause)?

Nik-09 replied:
Updated

seanlip commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts:
```diff
@@ -86,7 +86,7 @@ export class DiagnosticTestPlayerEngineService {
       this._diagnosticTestCurrentTopicStatusModel = (
         new DiagnosticTestCurrentTopicStatusModel(response));
       this._currentSkillId = (
```
Either add a comment to explain why it's guaranteed that getPendingSkillIds() will return a nonempty list, or handle the empty-list case correctly.

Nik-09 replied:
Added code comment

Nik-09 replied:
Added code comment

seanlip commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts:
```diff
@@ -132,7 +132,9 @@ export class DiagnosticTestPlayerEngineService {
           this._currentSkillId = (
-            this._diagnosticTestCurrentTopicStatusModel.getNextSkill());
+            this._diagnosticTestCurrentTopicStatusModel
+              .getPendingSkillIds()[0]
```
Ditto.

seanlip commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts:
```diff
@@ -146,7 +148,7 @@ export class DiagnosticTestPlayerEngineService {
       if (!this._diagnosticTestCurrentTopicStatusModel.isLifelineConsumed()) {
         this._currentSkillId = (
-          this._diagnosticTestCurrentTopicStatusModel.getNextSkill());
+          this._diagnosticTestCurrentTopicStatusModel.getPendingSkillIds()[0]);
```
Ditto.

Nik-09 replied:
Added

seanlip commented on core/templates/pages/exploration-player-page/services/exploration-player-state.service.ts:
```diff
@@ -332,11 +361,55 @@ export class ExplorationPlayerStateService {
+    ) {
+      return false;
+    } else {
+      throw new Error('Invalid mode received.');
```
Maybe `'Invalid mode received: ...` (i.e. add the invalid mode name to the error message).

seanlip commented on core/templates/pages/exploration-player-page/services/exploration-player-state.service.ts:
```diff
@@ -332,11 +361,55 @@ export class ExplorationPlayerStateService {
+    ExplorationPlayerConstants.EXPLORATION_MODE.EXPLORATION;
+  }
+
+  isInOtherMode(): boolean {
```
What's this for -- I thought you were removing this construct, per one of your other comments?

Nik-09 replied:
Currently, the exploration player state has different modes, among which there is a mode with the name “other”, this is naming is confusing and can be updated with a more meaningful name. 

Issue created: https://github.com/oppia/oppia/issues/16582

seanlip commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -323,6 +325,11 @@ var AdminPage = function() {
   };
 
+  this.createDummyClassroom = async function() {
+    await action.click('Activities Tab', activitiesTab);
```
Are you sure the activities tab actually exists on prod mode? The Activities tab in the admin page is hidden in that mode, you might need to create the necessary constructs manually (see other tests).

seanlip replied:
(I think this might be the cause of the failure here: https://github.com/oppia/oppia/actions/runs/3500183810/jobs/5862586972 )

Nik-09 replied:
Updated

seanlip reviewed (changes requested):
Hi @Nik-09, took a full pass and left some inline comments. Couple of questions:

- Please provide videos of the exploration player working correctly, the editor preview mode working correctly, and the regular question player working correctly.
- Are you running the e2e tests locally and watching them run? Do they work on your machine (using the same command as the one in .github/workflows/ in the github repo)?

Nik-09 committed:
Adds internationalized texts

Nik-09 committed:
Fixes typo

ghost commented:
> 

@seanlip @Nik-09 I don't know where the "Go Back" would take the user. Back to the diagnostic test? If it exits the test, I think "Exit Test" would be a more appropriate label.

seanlip commented:
Ah ok, thanks @findimue, I agree with the updated naming. 

Is the location of the button OK?

ghost commented:
I think the location is OK.

On Sat, Nov 19, 2022 at 2:22 PM Sean Lip ***@***.***> wrote:

> Ah ok, thanks @findimue <https://github.com/findimue>, I agree with the
> updated naming.
>
> Is the location of the button OK?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/16489#issuecomment-1320982097>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIFUUFCJVUESBDAJEZKCH4LWJFHMDANCNFSM6AAAAAARXNMCGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_session

Nik-09 committed:
Adds test code.

Nik-09 committed:
Diagnostic test e2e test

Nik-09 committed:
Removes unwanted code.

Nik-09 committed:
Updates varible name.

Nik-09 committed:
Adds code comment

Nik-09 committed:
Adds internationalized text

Nik-09 committed:
Fixes e2e test

Nik-09 committed:
Adds todo comment

Nik-09 committed:
Fixes frontend test

Nik-09 commented:
> Also @Nik-09 In the `two.mp3` video, the question text still doesn't seem to clear properly when a new question is shown. Have you taken a look into that?

Yes, I have taken a look into it --  this basically arises when we have dummy data generated from the admin activities tab and I think it is non-trivial to handle in this PR. The problem is the same for the existing practice session (as we have discussed earlier). Currently, I have not initiated the discussion with the Angular team because I am focusing on this PR. After this, I will create a design doc and then share it with the different members for review. 
Thank you

Nik-09 committed:
Make skip button variable available from the top level

Nik-09 reviewed (commented):
Hi @seanlip please take a look, I have addressed the comments please take a look.

Thank you

Nik-09 committed:
Adds feedback flag for embedded explorations

seanlip commented on core/templates/pages/exploration-player-page/layout-directives/progress-nav.component.spec.ts:
```diff
@@ -254,41 +254,6 @@ describe('Progress nav component', () => {
     expect(componentInstance.skipQuestion.emit).toHaveBeenCalled();
   });
 
-  it(
```
Why was this deleted?

Nik-09 replied:
Initially, the `skipButtonIsShown` variable is set in this component, and based on that this test block was written. But I have updated it to send this value from the top hierarchical component. So this test block is not needed anymore.
Thank you

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -481,17 +482,12 @@ export class ConversationSkinComponent {
-        this.enableCheckpointCelebrationModal = true;
+        this.feedbackIsEnabled = false;
+        this.learnerCanOnlyAttemptQuestionOnce = true;
+        this.lastInputOutputPairIsOnlyShown = true;
```
Should be onlyLastInputResponsePairIsShown -- what you have here means something different (it suggests that this is controlling whether the only thing we do with the last input output pair is to show it, and whether or not to do other things with it).

If you prefer, you can flip it and say something like "inputOutputHistoryIsShown" -- that might actually be clearer.

Nik-09 replied:
Updated

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -1299,7 +1295,7 @@ export class ConversationSkinComponent {
           }
         }
         if (
```
Add a line break above this, because it's not actually connected to the previous if clause (correct?)

Nik-09 replied:
Added

seanlip commented on core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts:
```diff
@@ -85,6 +85,11 @@ export class DiagnosticTestPlayerEngineService {
         new DiagnosticTestCurrentTopicStatusModel(response));
+
+      // The diagnostic test current topic status model is initialized in the
+      // above step, so there always be at least one skill ID in the pending
```
there --> there will

Ditto elsewhere.

Nik-09 replied:
Updated

seanlip commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -16,60 +16,137 @@
+    await users.logout();
   });
 
   it('should be able to start diagnostic test', async function() {
```
Shouldn't you have at least one test that reaches the "completion" modal and verify that recommendations are actually shown to the user? Otherwise your tests here have no expectations.

Nik-09 replied:
Updated the tests

seanlip commented on core/tests/webdriverio_utils/ClassroomPage.js:
```diff
@@ -40,6 +42,10 @@ var ClassroomPage = function() {
     }
   };
+
+  this.takeDiagnosticTest = async function() {
```
This is misnamed, it suggests that the learner has completed the whole test. I suggest calling it "launchDiagnosticTestPage()" or similar.

Nik-09 replied:
Updated

seanlip reviewed (changes requested):
Thanks @Nik-09! Just small comments left. 

Please also respond to @findimue's comments and explain how you have handled each of them.

U8NWXD commented on core/tests/webdriverio_utils/ClassroomPage.js:
```diff
@@ -18,9 +18,11 @@
 
 var ClassroomPage = function() {
   var topicSummaryTile = $('.e2e-test-topic-summary-tile');
+  var takeDiagnosticTest = $('.e2e-test-take-diagnostic-test');
```
Please name element selectors as nouns, e.g. "launchDiagnosticTestPageButton"

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_utils/ClassroomPage.js:
```diff
@@ -40,6 +42,10 @@ var ClassroomPage = function() {
   };
+
+  this.takeDiagnosticTest = async function() {
+    await action.click('Take diagnostic test', takeDiagnosticTest);
```
Like the selector name, the element description should be phrased as a noun

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_utils/ExplorationPlayerPage.js:
```diff
@@ -74,6 +74,7 @@ var ExplorationPlayerPage = function() {
   var viewSolutionButton = $('.e2e-test-view-solution');
   var voiceoverLanguageSelector = $('.e2e-test-audio-lang-select');
   var waitingForResponseElem = $('.e2e-test-input-response-loading-dots');
+  var skipQuestion = $('.e2e-test-skip-question-button');
```
ditto: name as noun

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_utils/ExplorationPlayerPage.js:
```diff
@@ -306,6 +307,10 @@ var ExplorationPlayerPage = function() {
   };
 
+  this.skipQuestion = async function() {
+    await action.click('skip question', skipQuestion);
```
ditto: name as noun

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_utils/DiagnosticTestPage.js:
```diff
@@ -0,0 +1,53 @@
+  var createNewClassroomButton = $('.e2e-test-create-new-classroom');
+
+  this.startDiagnosticTest = async function() {
+    await action.click('Start diagnostic test', startDiagnosticTestButton);
```
ditto: name as noun, here and below

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,152 @@
+      new TopicEditorPage.TopicEditorPage());
+
+    await users.createAndLoginCurriculumAdminUser(
+      'creator@classroomPage.com', 'creatorClassroomPage');
```
Instead of `classroomPage.com` and `ClassroomPage`, this should refer to the diagnostic test page. This procedure is documented on the [E2E wiki page](https://github.com/oppia/oppia/wiki/End-to-End-Tests#independence)

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,152 @@
+    var handle = await browser.getWindowHandle();
+    await topicsAndSkillsDashboardPage.get();
+    await topicsAndSkillsDashboardPage.createTopic(
+      'Topic 1', 'topic-one', 'Description', false);
```
Topics should be named uniquely, see https://github.com/oppia/oppia/wiki/End-to-End-Tests#independence

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,152 @@
+      topicsAndSkillsDashboardPage.createSkillWithDescriptionAndExplanation(
+        'Skill 1', 'Concept card explanation', false));
+    await skillEditorPage.addRubricExplanationForDifficulty(
+      'Easy', 'Second explanation for easy difficulty.');
```
Why is this the "second" explanation? Where's the first?

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,152 @@
+    (
+      await
+      topicsAndSkillsDashboardPage.createSkillWithDescriptionAndExplanation(
+        'Skill 1', 'Concept card explanation', false));
```
Skills should also have globally unique names

Nik-09 replied:
Updated the name

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,152 @@
+    await classroomPage.takeDiagnosticTest();
+    await waitFor.pageToFullyLoad();
+    await diagnosticTestPage.startDiagnosticTest();
+    await explorationPlayerPage.submitAnswer('TextInput', 'correct');
```
What's supposed to happen after an answer is submitted? That should be asserted.

Also, this only tests the "happy path" where the user gets the answer correct. Please also test incorrect answers

Nik-09 replied:
Added more test cases.

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,152 @@
+      await classroomPage.takeDiagnosticTest();
+      await waitFor.pageToFullyLoad();
+      await diagnosticTestPage.startDiagnosticTest();
+      await explorationPlayerPage.skipQuestion();
+      await explorationPlayerPage.skipQuestion();
```
Please assert that skipping questions actually works correctly. Also, is it worth ensuring that after a question is skipped, the user can still submit the next question? Finally, where are we testing that the user can get topic recommendations as mentioned in the test name?

Nik-09 replied:
Added more test cases

U8NWXD reviewed (changes requested):
Please also make very sure that this PR doesn't introduce any E2E flakes. If it does, we'll have to revert it to avoid letting develop become flaky again

Nik-09 committed:
Modifies code comment and variable name

Nik-09 committed:
Adds e2e test

Nik-09 committed:
Adds info text for mcq questions

Nik-09 reviewed (commented):
Hi Sean, I have addressed the comments and updated the screenshot (containing I don’t know option and a note for mcq questions).

I have added the e2e test for the diagnostic test which expects the number of recommended topics. I have tried to run the e2e test locally, but I am not able to run them. The procedure to log in as curriculum admin is failing every time.

I am trying to debug them by running different tests, maybe on different branches, to identify the issue, but can you please take a look into them & can guide me on which line may cause a problem?

Thank you.

Nik-09 commented:
@findimue I have updated the case for MCQ type of questions, PTAL.
Thank you

Nik-09 committed:
Video recording + waitFor in the diagnostic test

Nik-09 committed:
Adds waitfor in the diagnostic test page

Nik-09 committed:
Fixes e2e test

Nik-09 committed:
Removes e2e github workflow for testing

Nik-09 committed:
Fix selector for e2e test

Nik-09 committed:
Adds webpack build

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into diagnostic_test_session

Nik-09 commented on .github/workflows/e2e_translation_classroom_and_core_features.yml:
```diff
@@ -46,20 +46,15 @@ jobs:
-      - name: Run Exploration Translation Tab E2E Test
+      - name: Run Diagnostic Test Player Page E2E Test
         if: ${{ steps.check-risk.outputs.is-low-risk != 0 }}
-        run: xvfb-run -a --server-args="-screen 0, 1285x1000x24" python -m scripts.run_e2e_tests --skip-build --skip-install --suite="explorationTranslationTab" --prod_env
```
These lines are removed to run the workflow faster in the CI. I will add them back after fixing the test.

ghost commented:
The message looks fine, but the placement of the "I Don't Know" button is
now too close to the bottom of the frame. It should have some padding.

On Mon, Nov 21, 2022 at 1:20 AM Nikhil ***@***.***> wrote:

> @findimue <https://github.com/findimue> I have updated the case for MCQ
> type of questions, PTAL.
> Thank you
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/16489#issuecomment-1321744844>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIFUUFHJXI2YOBA7KNCIDJLWJM5EVANCNFSM6AAAAAARXNMCGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Nik-09 committed:
Trial e2e test fix

seanlip commented:
> I have added the e2e test for the diagnostic test which expects the number of recommended topics. I have tried to run the e2e test locally, but I am not able to run them. The procedure to log in as curriculum admin is failing every time.

> I am trying to debug them by running different tests, maybe on different branches, to identify the issue, but can you please take a look into them & can guide me on which line may cause a problem?

Hi @Nik-09, I tried looking into this but I don't have enough info to help you. When you say the procedure is failing, what is the error message in GitHub Actions exactly? It seems like the curriculum admin method was already there from before so I'm not sure why it would fail now unless there was some issue in regular production usage.

I'd need more details, e.g. a debugging doc explaining what you've tried and what exactly is failing, in order to be helpful -- I'd need to be able to at least figure out how to reproduce the problem on my end.

seanlip commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.spec.ts:
```diff
@@ -2298,7 +2298,7 @@ describe('Conversation skin component', () => {
       expect(componentInstance.feedbackIsEnabled).toBeTrue();
       expect(componentInstance.learnerCanOnlyAttemptQuestionOnce).toBeFalse();
-      expect(componentInstance.lastInputOutputPairIsOnlyShown).toBeFalse();
+      expect(componentInstance.inputOutputHistoryIsShown).toBeFalse();
```
Shouldn't the "sense" of inputOutputHistoryIsShown be opposite to that of lastInputOutputPairIsOnlyShown? I.e. shouldn't true <-> false?

inputOutputHistoryIsShown means we show the full conversation history.
lastInputOutputPairIsOnlyShown means we show only the last input-output pair.
They are opposites of each other. Please fix throughout.

Nik-09 replied:
Updated

seanlip reviewed (changes requested):
Just one (major) comment, otherwise looks good from my end!

Nik-09 committed:
Trial2 e2e test fix

Nik-09 committed:
Addresses review comments

Nik-09 committed:
Adds github workflow cmds back

dchen97 commented:
@findimue - What do you think about the varying button styles throughout the feature? I feel like the "Start test" button is especially different but I think across the feature it's very clear the buttons all have different styles with font, text size, and padding. Should we try to make them consistent?

dchen97 commented:
@Nik-09 - Out of curiosity - why isn't there a submission button for the multiple choice question like there is in the lesson player? Feels odd that this is inconsistent especially given users cannot go back to correct an incorrect answer

Nik-09 committed:
Fixes e2e test

Nik-09 committed:
Fixes e2e test

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Disables video recording

ghost commented:
I agree that the buttons should be more consistent.

> On Nov 21, 2022, at 6:22 PM, Diana ***@***.***> wrote:
> 
> ﻿
> @findimue - What do you think about the varying button styles throughout the feature? I feel like the "Start test" button is especially different but I think across the feature it's very clear the buttons all have different styles with font, text size, and padding. Should we try to make them consistent?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

Nik-09 committed:
Fixes lint issues

Nik-09 commented:
> @Nik-09 - Out of curiosity - why isn't there a submission button for the multiple choice question like there is in the lesson player? Feels odd that this is inconsistent especially given users cannot go back to correct an incorrect answer

Hmm not sure why this behavior is present i.e., not providing submit button in the MCQ type of questions. The submit buttons for MCQ questions are also not present for practice sessions -- 
![image](https://user-images.githubusercontent.com/57531197/203251173-e49f8981-6130-4fac-803b-d902859a72c2.png)


Also, I am updating the button styles for making them consistent throughout the feature.
Thank you

Nik-09 commented:
> Just one (major) comment, otherwise looks good from my end!

Hi, @seanlip I have addressed the comments and fixed the E2E test, now working on fixing merge conflicts and making the buttons more consistent.

Please take a look once you are free.

Thank you

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Updates buttons css

Nik-09 committed:
fixes typescript check

Nik-09 commented:
@dchen97 I have updated the UI screenshots, PTAL once you are free.
Thank you

Nik-09 commented:
@U8NWXD I have addressed the comments, PTAL once you are free.
Thank you

seanlip reviewed (approved):
Thanks @Nik-09! LGTM from my end.

U8NWXD commented on core/tests/webdriverio_utils/ClassroomPage.js:
```diff
@@ -40,6 +42,15 @@ var ClassroomPage = function() {
   };
+
+  this.launchDiagnosticTestPage = async function() {
+    await waitFor.elementToBeClickable(
+      launchDiagnosticTestPageButton,
+      'Launch diagnostic test button taking longer than expected'
+    );
```
You shouldn't need this. `action.click` includes a `waitFor.elementToBeClickable()` call already

Nik-09 replied:
Removed the unnecessary checks. Thank you for the guidance.

U8NWXD commented on core/tests/webdriverio_utils/ExplorationPlayerPage.js:
```diff
@@ -306,6 +307,14 @@ var ExplorationPlayerPage = function() {
   };
 
+  this.skipQuestion = async function() {
+    await waitFor.elementToBeClickable(
+      skipQuestionButton,
+      'Skip question button taking longer than expected'
+    );
```
ditto: unnecessary

Nik-09 replied:
Removed

U8NWXD commented on core/tests/webdriverio_utils/DiagnosticTestPage.js:
```diff
@@ -0,0 +1,151 @@
+  var classroomTileContainerSelector = $('.e2e-test-classroom-tile-container');
+
+  this.startDiagnosticTest = async function() {
+    await waitFor.pageToFullyLoad();
```
This call should go at the end of whatever function triggers a switch to a new page instead

Nik-09 replied:
Removed from here

U8NWXD commented on core/tests/webdriverio_utils/DiagnosticTestPage.js:
```diff
@@ -0,0 +1,151 @@
+
+  this.startDiagnosticTest = async function() {
+    await waitFor.pageToFullyLoad();
+    await waitFor.elementToBeClickable(
+      startDiagnosticTestButton,
+      'Start diagnostic test button taking too long to be clickable'
+    );
```
ditto: unnecessary. I'm going to stop commenting on all these cases, but please resolve the others like this too

Nik-09 replied:
Removed

U8NWXD commented on core/tests/webdriverio_utils/DiagnosticTestPage.js:
```diff
@@ -0,0 +1,151 @@
+      'Add new classroom button taking too long to be clickable'
+    );
+    await action.click(
+      'Add new classroom config data', addNewClassroomButton);
```
This should be named as a noun. Note that I flagged this issue earlier in https://github.com/oppia/oppia/pull/16489#discussion_r1027508168. When I said "here and below" in that comment, I meant cases like this one. ditto elsewhere

Nik-09 replied:
Updated

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,173 @@
+    var handle = await browser.getWindowHandle();
+    await topicsAndSkillsDashboardPage.get();
+    await topicsAndSkillsDashboardPage.createTopic(
+      'Addition', 'add', 'Addition of numbers', false);
```
This topic still needs to be named uniquely. Ditto elsewhere for topics and skills

Nik-09 replied:
Renamed: `Diagnostic test addition`

U8NWXD commented on core/tests/webdriverio_desktop/diagnosticTestPage.js:
```diff
@@ -0,0 +1,173 @@
+    await topicsAndSkillsDashboardPage.get();
+
+    (
+      await
+      topicsAndSkillsDashboardPage.createSkillWithDescriptionAndExplanation(
+        'Skill Addition', 'Concept card explanation', false));
```
This skill name is still not reliably unique. Please see the guidelines at https://github.com/oppia/oppia/wiki/End-to-End-Tests#independence

Nik-09 replied:
Renamed: `Diagnostic test skill`

Nik-09 committed:
Removes unnecessary clickable checks

Nik-09 reviewed (commented):
Hi, @U8NWXD thank you for the review, I have addressed the comments.
Please take a look at the nearest time when you are available.
Thank you.

dchen97 commented:
Thanks @Nik-09 ! This LGTM

U8NWXD commented on core/tests/webdriverio_utils/DiagnosticTestPage.js:
```diff
@@ -0,0 +1,127 @@
+    await waitFor.visibilityOf(
+      newClassrooomNameInput,
+      'New classroom name input is not visible'
+    );
```
This is also unnecessary because `setValue` includes an `action.click` call: 
https://github.com/oppia/oppia/blob/e915b3dccfcc629800b6fa99ccc2817e4d7ae562/core/tests/webdriverio_utils/action.js#L95-L101

As mentioned in the previous review (https://github.com/oppia/oppia/pull/16489#discussion_r1029886501) please remove the redundant waitFor calls from elsewhere too, not just here

Nik-09 replied:
Thank you @U8NWXD for guiding me on this, I was unaware of this.
I have removed the redundant check for visibility or clickability before the setValue or click events.

Please take a look.

Thank you

Nik-09 committed:
Removes redundant visbility checks from e2e test

Nik-09 committed:
Merge branch 'develop' into diagnostic_test_session
